src/crimson/os/seastore/nvmedevice/nvmedevice.cc

```diff
@@ -0,0 +1,227 @@
+}
+
+open_ertr::future<>
+NormalNBD::open(const std::string &in_path, seastar::open_flags mode) {
+  /* Open with posix fd for pass generic NVMe commands */
+  fd = seastar::file_desc::open(in_path, (int)mode);
+  identify_controller_data_t controller_data = {0, };
+  return identify_controller(controller_data).safe_then(
+      [this, controller_data, in_path, mode]() {
+      protocol_version = controller_data.version;
+      logger().debug("nvme protocol {}.{} {}",
+          (uint32_t)protocol_version.major_ver,
+          (uint32_t)protocol_version.minor_ver,
+          (uint32_t)protocol_version.tertiary_ver);
+
+      /*
+       * Multi Stream Write
+       *
+       * When NVMe SSD supports multi stream functionality, it marks oacs bit of
+       * identify_controller_data structure (from NVMe Specification 1.4).
+       * If oacs field is true, NormalNBD class opens device file multiple times
+       * with different stream IDs. When user calls write() with stream argument,
+       * NormalNBD finds pre-opened FD with stream ID and submit write IO to the
+       * found FD.
+       */
+      support_multistream = controller_data.oacs.support_directives;
+      if (support_multistream) {
+        write_life_max = 6;
+      }
+
+      open_for_io(in_path, mode);
+
+      /* PWG and PWA are supported from NVMe 1.4 */
+      if (protocol_version.major_ver >= 1 && protocol_version.minor_ver >= 4) {
+        identify_namespace_data_t namespace_data = {0, };
+        identify_namespace(namespace_data).safe_then([this, namespace_data]() {
+            /* Revise 0-based value */
+            write_granularity = namespace_data.npwg + 1;
+            write_alignment = namespace_data.npwa + 1;
+            });
+      }
+      return seastar::now();
+      }).handle_error(
+        /* If device does not support ioctl, just open without stream */
+        crimson::ct_error::input_output_error::handle([this, in_path, mode](auto) {
+          open_for_io(in_path, mode);
+          return seastar::now();
+        }));
+}
+
+void
+NormalNBD::open_for_io(const std::string &in_path, seastar::open_flags mode)
+{
+  ctx.resize(write_life_max);
+  for (uint32_t i = 0; i < write_life_max; i++) {
+    stream_fd.push_back(seastar::file_desc::open(in_path, (int)mode));
+    if (i != write_life_not_set) {
+      int posix_fd = stream_fd[i].get();
+      fcntl(posix_fd, F_SET_FILE_RW_HINT, &i);
+    }
+
+    io_setup(MAX_EVENTS, &ctx[i]);
+  }
+  completion_poller = std::thread(poll_completion, &ctx, &exit);
+}
+
+write_ertr::future<>
+NormalNBD::write(
+  uint64_t offset,
+  bufferptr &bptr,
+  uint16_t stream) {
+  logger().debug(
+      "block: do_write offset {} len {}",
+      offset,
+      bptr.length());
+  io_context_t io_context = io_context_t();
+  io_prep_pwrite(
+      &io_context.cb,
+      stream_fd[stream].get(),
+      bptr.c_str(),
+      bptr.length(),
+      offset);
+  iocb* cb_ptr[1] = {&io_context.cb};
+  io_submit(ctx[stream], 1, cb_ptr);
```

Comment:
You can't make calls like this from the seastar reactor.  Please look into properly integrating this into seastar.

Reply:
We use posix fd because seastar::file class does not support ioctl, and to support stream. Is there right way to integrate this into seastar?

Reply:
You'll need to coordinate with the seastar upstream.